### PR TITLE
Users/mdakbar/GitHub release task change log 2.0

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

**Comment by Ajay-MS** [comment]
Create PR on behalf of @mdmdakbar

File: Tasks/GithubReleaseV0/README.md [hunk 1]
```diff
@@ -1 +1,45 @@
...
+## Overview
+
+The GitHub Releases task can be used to create/edit/discard a GitHub release directly from your CI/CD pipeline. This task  works on cross platform Azure Pipeline agents running Windows, Linux or Mac and uses GitHub APIs to manage GitHub releases. 
+
+## Contact Information
+
+Please report a problem at [Developer Community Forum](https://developercommunity.visualstudio.com/spaces/21/index.html) if you are facing problems in making this task work.  You can also share feedback about the task like, what more functionality should be added to the task, what other tasks you would like to have, at the same place.
+
+## Pre-requisites for the task
+The following pre-requisites required for the task to work properly:
```
Comment (prativen) [inline Tasks/GithubReleaseV0/README.md]: are required

File: Tasks/GithubReleaseV0/README.md [hunk 2]
```diff
@@ -1 +1,45 @@
...
+
+#### GitHub Service Connection
+In order to perform operations on the GitHub repository, the task needs a GitHub service connection with adequate permission. If it does not exist already, you can [create a github service connection](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/service-endpoints?view=vsts#sep-github) in for your azure pipelines project. Once the service connection is created, all you  need is the name of the service connection in this task.
```
Comment (prativen) [inline Tasks/GithubReleaseV0/README.md]: in your azure pipelines project

File: Tasks/GithubReleaseV0/README.md [hunk 3]
```diff
@@ -1 +1,45 @@
...
+
+* **Target:** This is the commit SHA for which the github release will be created. By default, the value is $(Build.SourceVersion) which corresponds to the commit for which the build was run. This field is ignored when using edit and discard actions.
+
+* **Tag:** This field allows you to configure the tag to be used for a release action. It can be done in 2 ways:
+                * Use tag associated with commit: This option is available only for create release action. If selected, the release will be created ~~using the latest tag~~ that is associated with this commit. 
+                * Use specified tag: When this option is used, the release will be created using the tag mentioned. You can also mention the tag using variables,  Eg. v\$(MajorVersion).\$(MinorVersion).\$(PatchVersion). For release edit and discard actions, it is mandatory to use this option. 
+                
+* **Release Title:** This is the title that will be used for release creation. If left empty, the tag name will be used as the release title.
```
Comment (prativen) [inline Tasks/GithubReleaseV0/README.md]: have we tested this?

File: Tasks/GithubReleaseV0/operations/Action.ts [hunk 4]
```diff
@@ -10,40 +10,67 @@ export class Action {
...
+            }
 
-            if (response.statusCode === 201) {
-                if (!tl.getBoolInput(Inputs.isDraft)) {
-                    console.log(tl.loc("CreateReleaseSuccess", response.body[GitHubAttributes.htmlUrl]));
+            if (!!releaseResponse && releaseResponse.statusCode === 200) {
+                throw new Error(tl.loc("ReleaseAlreadyExists", tag));
+            }
+            else {
+                tl.debug("Creating draft release");
```
Comment (prativen) [inline Tasks/GithubReleaseV0/operations/Action.ts]: didnt get why else means draft?

File: Tasks/GithubReleaseV0/operations/Action.ts [hunk 5]
```diff
@@ -10,40 +10,67 @@ export class Action {
...
+                let response: WebResponse = await Release.createRelease(githubEndpoint, repositoryName, target, tag, releaseTitle, releaseNote, true, isPrerelease);
+                tl.debug("Create release response:\n" + JSON.stringify(response, null, 2));
+
+                if (response.statusCode === 201) {
```
Comment (prativen) [inline Tasks/GithubReleaseV0/operations/Action.ts]: this does not feel readable. too much logic and nesting

File: Tasks/GithubReleaseV0/operations/Utility.ts [hunk 6]
```diff
@@ -72,12 +72,21 @@ export class Utility {
         }
         tl.debug("ReleaseNote:\n" + releaseNote);
 
+        if (!releaseNote) {
+            releaseNote = "";
+        }
+
+        // Append commits and issues to release note.
+        if (changeLog){
+            releaseNote = releaseNote + changeLog;
```
Comment (prativen) [inline Tasks/GithubReleaseV0/operations/Utility.ts]: some newlines and header before appending changeLog?

File: Tasks/GithubReleaseV0/operations/Utility.ts [hunk 7]
```diff
@@ -87,7 +96,81 @@ export class Utility {
         return undefined;
     }
 
-    private static _tagRef: string = "refs/tags/";
+    public static parseHTTPHeaderLink(headerLink: string) {
```
Comment (prativen) [inline Tasks/GithubReleaseV0/operations/Utility.ts]: return type?

Comment (prativen) [reply Tasks/GithubReleaseV0/operations/Utility.ts]: comments? not readable

File: Tasks/GithubReleaseV0/operations/webClient.ts [hunk 8]
```diff
@@ -33,6 +33,11 @@ export class WebResponse {
     public body: any;
 }
 
+export class HTTPAttributes {
```
Comment (prativen) [inline Tasks/GithubReleaseV0/operations/webClient.ts]: are these Github or http attributes?

File: Tasks/GithubReleaseV0/operations/Utility.ts [hunk 9]
```diff
@@ -55,8 +55,8 @@ export class Utility {
         }
     }
 
-    public static getReleaseNote(releaseNotesSelection: string, releaseNotesFile: any, releaseNoteInput: string): string {
-        let releaseNote: string = undefined;
+    public static getReleaseNote(releaseNotesSelection: string, releaseNotesFile: any, releaseNoteInput: string, changeLog: string): string {
+        let releaseNote: string = "";
```
Comment (prativen) [inline Tasks/GithubReleaseV0/operations/Utility.ts]: where is changelog being generated?

File: Tasks/GithubReleaseV0/task.json [hunk 10]
```diff
@@ -3,9 +3,8 @@
     "name": "GithubRelease",
     "friendlyName": "Github Release",
     "description": "Create, edit or discard the github release",
-    "helpMarkDown": "[More Information](Todo: mdakbar: Link to github release documentation)",
+    "helpMarkDown": "[More Information](https://github.com/Microsoft/azure-pipelines-tasks/blob/master/Tasks/GithubReleaseV0/README.md)",
```
Comment (rajatagrawal-dev) [inline Tasks/GithubReleaseV0/task.json]: Get an aka.ms URL

File: Tasks/GithubReleaseV0/task.json [hunk 11]
```diff
@@ -26,15 +25,15 @@
             "label": "GitHub Connection",
             "defaultValue": "",
             "required": true,
-            "helpMarkDown": "Select the service endpoint for your GitHub connection. To create one, click the Manage link and create a new service endpoint."
+            "helpMarkDown": "Select the service endpoint for your GitHub connection. Click the 'Manage' link to create a new service endpoint. To know more about service connections [click here.](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/service-endpoints?view=vsts#sep-github)"
```
Comment (rajatagrawal-dev) [inline Tasks/GithubReleaseV0/task.json]: Get an aka.ms URL

File: Tasks/GithubReleaseV0/task.json (hunk 10 above)
Comment (rajatagrawal-dev) [inline Tasks/GithubReleaseV0/task.json]: Also add the releaseNotes field

**Comment by rajatagrawal-dev** [comment]
Check if you also need to add this task here https://github.com/Microsoft/azure-pipelines-tasks/blob/master/make-options.json

**Pull request merged**